### PR TITLE
Implement password reset functionality

### LINKED ISSUE #295: Implement Change Password

### PR DESCRIPTION
**Pull request by Pdzly**
Added password reset support including token generation, email notifications and password update. This involves updates in JwtUtil, JwtFilter, PersonService classes and creation of PasswordReset entity for storing reset token. PasswordResetService was introduced to handle password-reset-related operations along with necessary events for handling password reset actions. Updated database migration script for the changes.

closes #295

### CODE REVIEW

File: src/main/java/com/sublinks/sublinksapi/api/lemmy/v3/authentication/JwtUtil.java [hunk 1]
```diff
@@ -24,7 +25,7 @@ public class JwtUtil implements Serializable {
   private static final long serialVersionUID = -2550185165626007488L;
   private final String secret;
 
-  public JwtUtil(@Value("${jwt.secret}") final String secret) {
+  public JwtUtil(@Value("${jwt.secret}") final String secret, UserDataService userDataService) {
```
Comment (jgrim) [inline src/main/java/com/sublinks/sublinksapi/api/lemmy/v3/authentication/JwtUtil.java:28]: Why this change?

**Review by jgrim — APPROVED** [review]

File: src/main/java/com/sublinks/sublinksapi/api/lemmy/v3/authentication/JwtUtil.java (hunk 1 above)
Comment (Pdzly) [reply src/main/java/com/sublinks/sublinksapi/api/lemmy/v3/authentication/JwtUtil.java:28]: Removed it, it was just an oversight from my testing.

File: src/main/java/com/sublinks/sublinksapi/api/lemmy/v3/authentication/JwtUtil.java [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package com.sublinks.sublinksapi.api.lemmy.v3.authentication;
 
 import com.sublinks.sublinksapi.person.dto.Person;
+import com.sublinks.sublinksapi.person.services.UserDataService;
```
Comment (jgrim) [inline src/main/java/com/sublinks/sublinksapi/api/lemmy/v3/authentication/JwtUtil.java:4]: This can be removed too

**Review by jgrim — CHANGES_REQUESTED** [review]

**Review by jgrim — APPROVED** [review]

File: src/main/java/com/sublinks/sublinksapi/person/services/PasswordResetService.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    return UUID.randomUUID().toString();
+  }
+
+  public PasswordReset createPasswordReset(Person person) {
+
+    PasswordReset passwordReset = new PasswordReset();
+    passwordReset.setPerson(person);
+    passwordReset.setUsed(false);
+    passwordReset.setCreatedAt(new Date());
+    passwordReset.setToken(generateToken());
```
Comment (DarkArc) [inline src/main/java/com/sublinks/sublinksapi/person/services/PasswordResetService.java:39]: This is a _bit_ of a nit, but I would recommend against using UUIDs for a password reset token.

The Java UUID generator does make an effort to prevent predictability but that not its primary objective: https://docs.oracle.com/javase/8/docs/api/java/util/UUID.html#randomUUID

A better approach would likely be to generated N number of bytes from SecureRandom

https://stackoverflow.com/a/56705603

Another strategy here would be cryptographically signing the generated ID (i.e., outright issue a scoped JWT)

https://github.com/sublinks/sublinks-api/blob/6afdd061aa698a9bb7a7412a0b5535511c746f0c/src/main/java/com/sublinks/sublinksapi/api/lemmy/v3/authentication/JwtUtil.java#L71

That would require that both the ID is predicted _and_ signed with the server's private key which _should_ be next to impossible to perform without a broader system compromise.
